File: runtime/fleet-mobile/modules/fleet-console-view/android/src/main/java/com/dotobokuri/fleet/mobile/RemoteConnection.kt
package com.dotobokuri.fleet.mobile

import android.os.Handler
import android.os.Looper
import android.webkit.CookieManager
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.URI
import java.net.URL
import java.security.MessageDigest
import java.security.cert.CertificateException
import java.security.cert.X509Certificate
import java.util.Date
import java.util.Locale
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import javax.net.ssl.HostnameVerifier
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLContext
import javax.net.ssl.SSLSession
import javax.net.ssl.SSLSocket
import javax.net.ssl.X509TrustManager

internal class RemoteConnection(
  private val cookies: FleetCookieStore = FleetCookieStore(),
) {
  fun preflight(target: PersistedTarget) {
    openPinnedSocket(target).use { }
  }

  fun openPinnedSocket(target: PersistedTarget): SSLSocket {
    val trust = StrictPinnedTrustManager(target)
    val context = SSLContext.getInstance("TLS").apply { init(null, arrayOf(trust), null) }
    val socket = context.socketFactory.createSocket() as SSLSocket
    try {
      socket.soTimeout = TIMEOUT_MS
      socket.enabledProtocols = socket.supportedProtocols.filter { it == "TLSv1.3" || it == "TLSv1.2" }.toTypedArray()
      socket.connect(InetSocketAddress(target.hostname, target.port), TIMEOUT_MS)
      val parameters = socket.sslParameters
      // A raw TLS socket sends no ALPN offer, so the upstream cannot select HTTP/2.
      parameters.endpointIdentificationAlgorithm = "HTTPS"
      if (!CertificatePolicy.isIpLiteral(target.hostname)) {
        parameters.serverNames = listOf(javax.net.ssl.SNIHostName(target.hostname))
      }
      socket.sslParameters = parameters
      socket.startHandshake()
      val certificate = socket.session.peerCertificates.firstOrNull() as? X509Certificate
        ?: throw ConnectionFailure("remote_link_unverified")
      CertificatePolicy.verifyLeaf(certificate, target)
      return socket
    } catch (failure: ConnectionFailure) {
      try { socket.close() } catch (_: Exception) { }
      throw failure
    } catch (failure: Exception) {
      try { socket.close() } catch (_: Exception) { }
      throw mappedTlsFailure(failure)
    }
  }

  fun join(target: PersistedTarget, token: String?, deviceName: String) {
    request(
      target = target,
      url = target.joinUrl,
      method = "POST",
      body = JSONObject().apply {
        if (token != null) put("token", token)
        put("device", deviceName)
      }.toString(),
    ).use { response ->
      when (response.code) {
        in 200..299 -> Unit
        401 -> throw ConnectionFailure(if (token == null) "remote_host_not_paired" else "remote_link_rejected")
        403 -> throw ConnectionFailure("remote_link_host_mismatch")
        409 -> throw ConnectionFailure(
          if (response.errorCode() == "paired_device_limit") "remote_link_device_limit" else "remote_link_control_held",
        )
        // The join guard rejects these before reading the body, so even a valid token is refused
        // until Retry-After passes; they are wait states, not identity failures.
        429 -> throw ConnectionFailure("remote_link_throttled", retryAfterSeconds = response.retryAfterSeconds())
        503 -> throw ConnectionFailure("remote_host_busy", retryAfterSeconds = response.retryAfterSeconds())
        else -> throw ConnectionFailure("remote_link_unverified")
      }
    }
  }

  fun verifyReachable(target: PersistedTarget) {
    request(target, target.readinessUrl, "GET", null).use { response ->
      if (response.code == 401) throw ConnectionFailure("remote_host_session_expired")
      if (response.code !in 200..299 || !response.isConsoleStatus()) {
        throw ConnectionFailure("remote_host_unavailable")
      }
    }
  }

  private fun request(target: PersistedTarget, url: String, method: String, body: String?): Response {
    requireExpectedAuthority(target, url)
    val connection = URL(url).openConnection() as? HttpsURLConnection
      ?: throw ConnectionFailure("remote_link_unverified")
    connection.instanceFollowRedirects = false
    connection.connectTimeout = TIMEOUT_MS
    connection.readTimeout = TIMEOUT_MS
    connection.requestMethod = method
    connection.sslSocketFactory = SSLContext.getInstance("TLS").apply {
      init(null, arrayOf(StrictPinnedTrustManager(target)), null)
    }.socketFactory
    connection.hostnameVerifier = StrictHostnameVerifier(target)
    connection.setRequestProperty("Accept", "application/json")
    // CookieManager is the only cookie store. Read at the last possible moment for every request.
    cookies.readRemote(url, target.port)?.let { connection.setRequestProperty("Cookie", it) }
    if (body != null) {
      connection.doOutput = true
      connection.setRequestProperty("Content-Type", "application/json")
      connection.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
    }
    val code = try {
      connection.responseCode
    } catch (failure: Exception) {
      connection.disconnect()
      throw mappedTlsFailure(failure)
    }
    if (code in 300..399) {
      connection.disconnect()
      throw ConnectionFailure("remote_link_redirect_refused")
    }
    try {
      cookies.applyRemoteAndAwait(url, target.port, setCookieHeaders(connection))
    } catch (failure: Exception) {
      connection.disconnect()
      throw ConnectionFailure("remote_link_cookie_failed", failure)
    }
    return Response(connection, code)
  }

  private fun requireExpectedAuthority(target: PersistedTarget, value: String) {
    val uri = try { URI(value) } catch (_: Exception) { throw ConnectionFailure("remote_link_unverified") }
    val port = if (uri.port == -1) 443 else uri.port
    if (uri.scheme != "https" || uri.rawUserInfo != null || uri.rawFragment != null ||
      !uri.host.equals(target.hostname, ignoreCase = true) || port != target.port) {
      throw ConnectionFailure("remote_link_host_mismatch")
    }
    val renderedHost = if (target.hostname.contains(':')) "[${target.hostname}]" else target.hostname
    val expectedOrigin = if (target.port == 443) "https://$renderedHost" else "https://$renderedHost:${target.port}"
    if (target.origin != expectedOrigin) throw ConnectionFailure("remote_link_host_mismatch")
  }

  private fun setCookieHeaders(connection: HttpsURLConnection): List<String> = connection.headerFields.entries
    .filter { it.key?.equals("Set-Cookie", ignoreCase = true) == true }
    .flatMap { it.value.orEmpty() }

  private fun mappedTlsFailure(failure: Exception): ConnectionFailure {
    var current: Throwable? = failure
    while (current != null) {
      when (current.message) {
        "certificate_pin_mismatch" -> return ConnectionFailure("remote_link_fingerprint_mismatch", failure)
        "certificate_hostname_mismatch",
        "certificate_expired",
        "certificate_not_yet_valid",
        "certificate_not_leaf",
        "missing_leaf_certificate" -> return ConnectionFailure("remote_link_unverified", failure)
      }
      current = current.cause
    }
    return ConnectionFailure("remote_link_unreachable", failure)
  }

  private class Response(private val connection: HttpsURLConnection, val code: Int) : AutoCloseable {
    private fun body(): String? {
      val stream = if (code >= HttpURLConnection.HTTP_BAD_REQUEST) connection.errorStream else connection.inputStream
      return try { stream?.bufferedReader()?.use { it.readText() } } catch (_: Exception) { null }
    }

    fun errorCode(): String? = try {
      body()?.let { JSONObject(it).optString("error").takeIf(String::isNotEmpty) }
    } catch (_: Exception) { null }

    fun retryAfterSeconds(): Int? = RetryAfterHeader.seconds(connection.getHeaderField("Retry-After"))

    fun isConsoleStatus(): Boolean = try {
      val value = JSONObject(body() ?: return false)
      value.getString("name").isNotEmpty() && value.getString("version").isNotEmpty()
    } catch (_: Exception) { false }

    override fun close() { connection.disconnect() }
  }

  private class StrictPinnedTrustManager(private val target: PersistedTarget) : X509TrustManager {
    override fun getAcceptedIssuers(): Array<X509Certificate> = emptyArray()
    override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?) {
      throw CertificateException("client_certificates_not_accepted")
    }
    override fun checkServerTrusted(chain: Array<out X509Certificate>?, authType: String?) {
      val leaf = chain?.firstOrNull() ?: throw CertificateException("missing_leaf_certificate")
      CertificatePolicy.verifyLeaf(leaf, target)
    }
  }

  private class StrictHostnameVerifier(private val target: PersistedTarget) : HostnameVerifier {
    override fun verify(hostname: String?, session: SSLSession?): Boolean {
      if (hostname == null || session == null || !hostname.equals(target.hostname, ignoreCase = true)) return false
      val leaf = session.peerCertificates.firstOrNull() as? X509Certificate ?: return false
      return try {
        CertificatePolicy.verifyLeaf(leaf, target)
        true
      } catch (_: Exception) {
        false
      }
    }
  }

  companion object {
    private const val TIMEOUT_MS = 8_000

    fun fingerprint(certificate: X509Certificate): String = MessageDigest.getInstance("SHA-256")
      .digest(certificate.encoded)
      .joinToString("") { "%02X".format(Locale.ROOT, it.toInt() and 0xff) }
  }
}

internal data class FleetCookieSnapshot(val values: Map<String, String>)

internal object FleetCookiePolicy {
  private val valuePattern = Regex("^[A-Za-z0-9_-]*$")

  fun requestHeader(header: String?, port: Int): String? = scoped(header, port)
    .also { require(it.values.all(valuePattern::matches)) }
    .entries
    .joinToString("; ") { (name, value) -> "$name=$value" }
    .takeIf(String::isNotEmpty)

  fun snapshot(header: String?, port: Int): FleetCookieSnapshot = FleetCookieSnapshot(
    scoped(header, port).also { require(it.values.all(valuePattern::matches)) },
  )

  fun restoreHeaders(snapshot: FleetCookieSnapshot, currentHeader: String?, port: Int): List<String> {
    val currentNames = scoped(currentHeader, port).keys
    val expired = (currentNames - snapshot.values.keys).map {
      "$it=; Max-Age=0; Secure; HttpOnly; SameSite=Strict; Path=/"
    }
    val restored = snapshot.values.map { (name, value) ->
      val maxAge = if (name == pairingName(port)) "; Max-Age=31536000" else ""
      "$name=$value$maxAge; Secure; HttpOnly; SameSite=Strict; Path=/"
    }
    return expired + restored
  }

  fun validateSetCookieHeaders(values: List<String>, port: Int): List<String> {
    val seen = mutableSetOf<String>()
    return values.map { value ->
      val parts = value.split(';').map(String::trim)
      val pair = parts.firstOrNull()?.split('=', limit = 2).orEmpty()
      require(pair.size == 2 && pair[0] in names(port) && seen.add(pair[0]) && valuePattern.matches(pair[1]))
      val attributes = linkedMapOf<String, String?>()
      parts.drop(1).forEach { raw ->
        val attribute = raw.split('=', limit = 2)
        val name = attribute[0].lowercase(Locale.ROOT)
        require(name in setOf("httponly", "secure", "samesite", "path", "max-age") && name !in attributes)
        attributes[name] = attribute.getOrNull(1)
      }
      require(attributes["httponly"] == null && "httponly" in attributes)
      require(attributes["secure"] == null && "secure" in attributes)
      require(attributes["samesite"]?.equals("Strict", ignoreCase = true) == true)
      require(attributes["path"] == "/")
      val maxAge = attributes["max-age"]
      require("max-age" !in attributes || maxAge?.toLongOrNull()?.let { it >= 0 } == true)
      require(maxAge == null || pair[0] == pairingName(port))
      require(pair[1].isNotEmpty() || maxAge == "0")
      buildString {
        append(pair[0]).append('=').append(pair[1])
        if (maxAge != null) append("; Max-Age=").append(maxAge)
        append("; Secure; HttpOnly; SameSite=Strict; Path=/")
      }
    }
  }

  private fun names(port: Int): Set<String> = setOf(sessionName(port), pairingName(port))
  private fun sessionName(port: Int): String = "fleet_console_session_$port"
  private fun pairingName(port: Int): String = "fleet_console_pairing_$port"

  private fun scoped(header: String?, port: Int): Map<String, String> {
    val pairs = parse(header).filter { it.first in names(port) }
    require(pairs.map { it.first }.distinct().size == pairs.size)
    return pairs.toMap()
  }

  private fun parse(header: String?): List<Pair<String, String>> = header.orEmpty().split(';').mapNotNull {
    val pair = it.trim().split('=', limit = 2)
    if (pair.size == 2 && pair[0].isNotEmpty()) pair[0] to pair[1] else null
  }
}

internal class FleetCookieStore(
  private val manager: CookieManager = CookieManager.getInstance(),
  private val main: Handler = Handler(Looper.getMainLooper()),
) {
  fun readRemote(url: String, port: Int): String? = FleetCookiePolicy.requestHeader(manager.getCookie(url), port)

  fun snapshot(url: String, port: Int): FleetCookieSnapshot = FleetCookiePolicy.snapshot(manager.getCookie(url), port)

  fun restoreAndAwait(url: String, port: Int, snapshot: FleetCookieSnapshot) {
    applyAndAwait(url, FleetCookiePolicy.restoreHeaders(snapshot, manager.getCookie(url), port))
  }

  fun applyRemoteAndAwait(url: String, port: Int, values: List<String>) {
    applyAndAwait(url, FleetCookiePolicy.validateSetCookieHeaders(values, port))
  }

  fun applyAndAwait(url: String, values: List<String>) {
    values.forEach { value ->
      val latch = CountDownLatch(1)
      val accepted = AtomicBoolean(false)
      main.post {
        manager.setCookie(url, value) { success ->
          accepted.set(success)
          latch.countDown()
        }
      }
      if (!latch.await(COOKIE_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
        throw IllegalStateException("cookie_callback_timeout")
      }
      if (!accepted.get()) throw IllegalStateException("cookie_rejected")
    }
    // Preserve header order by waiting for each callback before applying the next cookie.
    manager.flush()
  }

  private companion object {
    const val COOKIE_TIMEOUT_MS = 8_000L
  }
}

internal object CertificatePolicy {
  fun verifyLeaf(certificate: X509Certificate, target: PersistedTarget, now: Date = Date()) {
    try {
      certificate.checkValidity(now)
    } catch (failure: java.security.cert.CertificateExpiredException) {
      throw CertificateException("certificate_expired", failure)
    } catch (failure: java.security.cert.CertificateNotYetValidException) {
      throw CertificateException("certificate_not_yet_valid", failure)
    }
    if (certificate.basicConstraints >= 0) throw CertificateException("certificate_not_leaf")
    if (!hasExactSubjectAlternativeName(certificate, target.hostname)) {
      throw CertificateException("certificate_hostname_mismatch")
    }
    if (RemoteConnection.fingerprint(certificate) != target.fingerprint) {
      throw CertificateException("certificate_pin_mismatch")
    }
  }

  fun hasExactSubjectAlternativeName(certificate: X509Certificate, hostname: String): Boolean {
    val expectedType = if (isIpLiteral(hostname)) 7 else 2
    return try {
      certificate.subjectAlternativeNames.orEmpty().any { name ->
        name.size >= 2 && name[0] == expectedType && when (expectedType) {
          7 -> sameIpAddress(hostname, name[1]?.toString().orEmpty())
          else -> name[1]?.toString()?.equals(hostname, ignoreCase = true) == true
        }
      }
    } catch (_: Exception) {
      false
    }
  }

  fun isIpLiteral(host: String): Boolean = host.contains(':') ||
    host.matches(Regex("^(?:0|[1-9]\\d{0,2})(?:\\.(?:0|[1-9]\\d{0,2})){3}$")) &&
      host.split('.').all { it.toIntOrNull() in 0..255 }

  private fun sameIpAddress(left: String, right: String): Boolean = try {
    InetAddress.getByName(left).address.contentEquals(InetAddress.getByName(right).address)
  } catch (_: Exception) {
    false
  }
}

/** Only the delta-seconds form is trusted; an HTTP-date or out-of-range value yields no countdown. */
internal object RetryAfterHeader {
  fun seconds(value: String?): Int? = value?.trim()?.toLongOrNull()?.takeIf { it in 0..86_400 }?.toInt()
}

internal class ConnectionFailure(
  val code: String,
  cause: Throwable? = null,
  val retryAfterSeconds: Int? = null,
) : Exception(code, cause)

File: runtime/fleet-mobile/modules/fleet-console-view/ios/Core/RemoteConnection.swift
import Foundation
import Network
import Security

// RemoteConnection.kt의 이식. 원격 콘솔로의 모든 연결은 시스템 CA를 쓰지 않는 leaf-only
// 핀 TLS다: TLS 1.2/1.3만, IP 리터럴엔 SNI 생략, ALPN 미제공(HTTP/2 회피), 8초 타임아웃,
// 리다이렉트 거부. join/verifyReachable은 Android의 HttpsURLConnection 대신 같은 핀 소켓
// 위에서 HttpCodec으로 HTTP/1.1을 직접 말한다 — TLS 경로가 하나로 유지되고 리다이렉트
// 미추적·헤더 통제가 코드로 보장된다.
//
// 실제 네트워크 동작(핸드셰이크·검증 블록·릴레이)은 기기에서만 검증된다
// [Unverified-on-device]; CI는 컴파일과 순수 로직(authority 검사·실패 매핑)만 증명한다.

public final class RemoteConnection {
  public static let timeoutMs = 8_000

  private let cookies: FleetCookieJar

  public init(cookies: FleetCookieJar) {
    self.cookies = cookies
  }

  // MARK: - Public surface (Kotlin과 동일)

  public func preflight(_ target: PersistedTarget) throws {
    let socket = try openPinnedSocket(target)
    socket.close()
  }

  public func join(_ target: PersistedTarget, token: String?, deviceName: String) throws {
    var body: [String: Any] = ["device": deviceName]
    if let token { body["token"] = token }
    let payload = try JSONSerialization.data(withJSONObject: body)
    let response = try request(target, path: "/api/v1/join", method: "POST", body: payload)
    switch response.code {
    case 200...299:
      return
    case 401:
      throw ConnectionFailure(token == nil ? "remote_host_not_paired" : "remote_link_rejected")
    case 403:
      throw ConnectionFailure("remote_link_host_mismatch")
    case 409:
      throw ConnectionFailure(response.errorCode() == "paired_device_limit" ? "remote_link_device_limit" : "remote_link_control_held")
    // join 가드는 바디를 읽기 전에 거절한다 — 유효한 토큰도 Retry-After가 지나기 전엔 거부되는
    // 대기 상태이지 identity 실패가 아니다.
    case 429:
      throw ConnectionFailure("remote_link_throttled", retryAfterSeconds: response.retryAfterSeconds())
    case 503:
      throw ConnectionFailure("remote_host_busy", retryAfterSeconds: response.retryAfterSeconds())
    default:
      throw ConnectionFailure("remote_link_unverified")
    }
  }

  public func verifyReachable(_ target: PersistedTarget) throws {
    let response = try request(target, path: "/api/v1/status", method: "GET", body: nil)
    if response.code == 401 { throw ConnectionFailure("remote_host_session_expired") }
    if !(200...299).contains(response.code) || !response.isConsoleStatus() {
      throw ConnectionFailure("remote_host_unavailable")
    }
  }

  /// 게이트웨이 릴레이가 업스트림으로 쓰는 새 핀 TLS 소켓.
  public func openPinnedSocket(_ target: PersistedTarget) throws -> PinnedSocket {
    try PinnedSocket.connect(target)
  }

  // MARK: - HTTP over the pinned socket

  private struct HttpResult {
    let code: Int
    let headers: Headers
    let body: Data

    func errorCode() -> String? {
      guard
        let object = (try? JSONSerialization.jsonObject(with: body)) as? [String: Any],
        let error = object["error"] as? String, !error.isEmpty
      else { return nil }
      return error
    }

    func retryAfterSeconds() -> Int? {
      RetryAfterHeader.seconds(headers.first("Retry-After"))
    }

    func isConsoleStatus() -> Bool {
      guard
        let object = (try? JSONSerialization.jsonObject(with: body)) as? [String: Any],
        let name = object["name"] as? String, !name.isEmpty,
        let version = object["version"] as? String, !version.isEmpty
      else { return false }
      return true
    }
  }

  private func request(_ target: PersistedTarget, path: String, method: String, body: Data?) throws -> HttpResult {
    try Self.requireExpectedAuthority(target, "\(target.origin)\(path)")
    let socket = try openPinnedSocket(target)
    defer { socket.close() }
    do {
      var headers = Headers()
      headers.append(name: "Host", value: target.authority)
      headers.append(name: "Accept", value: "application/json")
      // 쿠키는 요청 직전에 읽는다 — jar가 유일한 저장소다.
      if let cookie = try cookies.readRemote(target.origin, target.port) {
        headers.append(name: "Cookie", value: cookie)
      }
      if let body {
        headers.append(name: "Content-Type", value: "application/json")
        headers.append(name: "Content-Length", value: String(body.count))
      }
      headers.append(name: "Connection", value: "close")
      try HttpCodec.writeHead(socket, startLine: "\(method) \(path) HTTP/1.1", headers: headers)
      if let body { try socket.write([UInt8](body)) }
      try socket.flush()

      let response = try HttpCodec.readResponse(socket, requestMethod: method)
      // 원격의 3xx는 어떤 경우에도 따라가지 않는다.
      if (300...399).contains(response.code) { throw ConnectionFailure("remote_link_redirect_refused") }

      let bodyData = try Self.readBody(socket, response)
      // Set-Cookie 반영 실패는 연결 실패다(fail-closed).
      let setCookies = response.headers.values("Set-Cookie")
      if !setCookies.isEmpty {
        do { try cookies.applyRemote(target.origin, target.port, setCookies) }
        catch { throw ConnectionFailure("remote_link_cookie_failed") }
      }
      return HttpResult(code: response.code, headers: response.headers, body: bodyData)
    } catch let failure as ConnectionFailure {
      throw failure
    } catch {
      throw Self.mappedTlsFailure(error)
    }
  }

  private static let maxResponseBodyBytes = 1 << 20 // join/status는 작은 JSON — 1MiB 캡

  private static func readBody(_ socket: PinnedSocket, _ response: HttpResponse) throws -> Data {
    let sink = ByteArrayOutput()
    switch response.bodyKind {
    case .none:
      break
    case .fixed(let length):
      if length > Int64(maxResponseBodyBytes) { throw ConnectionFailure("remote_host_unavailable") }
      try HttpCodec.copyExactly(socket, sink, length: length)
    case .chunked:
      try HttpCodec.copyChunked(socket, sink, decode: true)
    case .untilClose:
      try HttpCodec.copyUntilEof(socket, sink, false)
    }
    if sink.bytes.count > maxResponseBodyBytes { throw ConnectionFailure("remote_host_unavailable") }
    return Data(sink.bytes)
  }

  // MARK: - Pure helpers (CI 테스트 대상)

  static func requireExpectedAuthority(_ target: PersistedTarget, _ value: String) throws {
    guard let uri = LocationUri.parse(value) else { throw ConnectionFailure("remote_link_unverified") }
    let port = uri.port == -1 ? 443 : uri.port
    guard
      uri.scheme == "https", uri.rawUserInfo == nil, uri.rawFragment == nil,
      let host = uri.host, host.caseInsensitiveCompare(target.hostname) == .orderedSame,
      port == target.port
    else { throw ConnectionFailure("remote_link_host_mismatch") }
    let renderedHost = target.hostname.contains(":") ? "[\(target.hostname)]" : target.hostname
    let expectedOrigin = target.port == 443 ? "https://\(renderedHost)" : "https://\(renderedHost):\(target.port)"
    if target.origin != expectedOrigin { throw ConnectionFailure("remote_link_host_mismatch") }
  }

  static func mappedTlsFailure(_ failure: Error) -> ConnectionFailure {
    if let certificate = failure as? CertificateError {
      switch certificate.code {
      case "certificate_pin_mismatch":
        return ConnectionFailure("remote_link_fingerprint_mismatch")
      case "certificate_hostname_mismatch", "certificate_expired", "certificate_not_yet_valid",
           "certificate_not_leaf", "missing_leaf_certificate":
        return ConnectionFailure("remote_link_unverified")
      default:
        break
      }
    }
    return ConnectionFailure("remote_link_unreachable")
  }
}

// NWConnection을 블로킹 IO로 감싼 핀 TLS 소켓. 게이트웨이 릴레이(ByteInput/ByteOutput)와
// HttpCodec이 같은 구현을 소비한다.
public final class PinnedSocket: ByteInput, ByteOutput {
  private let connection: NWConnection
  private let queue = DispatchQueue(label: "fleet-mobile-pinned-socket")
  private var readTimeoutMs: Int? = RemoteConnection.timeoutMs
  // 핸드셰이크 검증 블록이 기록하는 인증서 실패 — 연결 실패 시 매핑에 쓴다.
  private let verifyFailure: FailureBox

  private final class FailureBox: @unchecked Sendable {
    private let lock = NSLock()
    private var value: CertificateError?
    func set(_ error: CertificateError) { lock.lock(); value = error; lock.unlock() }
    func get() -> CertificateError? { lock.lock(); defer { lock.unlock() }; return value }
  }

  static func connect(_ target: PersistedTarget) throws -> PinnedSocket {
    let tls = NWProtocolTLS.Options()
    let options = tls.securityProtocolOptions
    sec_protocol_options_set_min_tls_protocol_version(options, .TLSv12)
    sec_protocol_options_set_max_tls_protocol_version(options, .TLSv13)
    // IP 리터럴엔 SNI를 보내지 않는다. 이름이면 SNI 설정(검증은 우리 verify 블록이 전담).
    if !CertificatePolicy.isIpLiteral(target.hostname) {
      sec_protocol_options_set_tls_server_name(options, target.hostname)
    }
    // ALPN을 제공하지 않으므로 업스트림은 HTTP/2를 선택할 수 없다.

    let socket = PinnedSocket(target: target, tls: tls)
    try socket.start(target)
    return socket
  }

  private init(target: PersistedTarget, tls: NWProtocolTLS.Options) {
    // 시스템 신뢰 평가를 우리 leaf-only 핀 정책으로 대체한다. (박스를 로컬로 먼저 만들어
    // self 초기화 완료 전에 클로저가 self를 캡처하지 않게 한다.)
    let box = FailureBox()
    verifyFailure = box
    sec_protocol_options_set_verify_block(tls.securityProtocolOptions, { _, trustRef, complete in
      let trust = sec_trust_copy_ref(trustRef).takeRetainedValue()
      guard
        let chain = SecTrustCopyCertificateChain(trust) as? [SecCertificate],
        let leaf = chain.first
      else {
        box.set(CertificateError("missing_leaf_certificate"))
        complete(false)
        return
      }
      do {
        try CertificatePolicy.verifyLeaf(leaf, target)
        complete(true)
      } catch let failure as CertificateError {
        box.set(failure)
        complete(false)
      } catch {
        box.set(CertificateError("certificate_not_leaf"))
        complete(false)
      }
    }, queue)

    let parameters = NWParameters(tls: tls)
    let host = NWEndpoint.Host(target.hostname)
    let port = NWEndpoint.Port(rawValue: UInt16(target.port)) ?? 443
    connection = NWConnection(host: host, port: port, using: parameters)
  }

  private func start(_ target: PersistedTarget) throws {
    let ready = DispatchSemaphore(value: 0)
    let state = FailureBox()
    connection.stateUpdateHandler = { [verifyFailure] update in
      switch update {
      case .ready:
        ready.signal()
      case .failed, .cancelled:
        // verify 블록이 기록한 실패가 있으면 그것이 원인이다.
        if state.get() == nil, let certificate = verifyFailure.get() { state.set(certificate) }
        else if state.get() == nil { state.set(CertificateError("__transport__")) }
        ready.signal()
      default:
        break
      }
    }
    connection.start(queue: queue)
    if ready.wait(timeout: .now() + .milliseconds(RemoteConnection.timeoutMs)) == .timedOut {
      connection.cancel()
      throw ConnectionFailure("remote_link_unreachable")
    }
    if let failure = state.get() {
      connection.cancel()
      if failure.code == "__transport__" { throw ConnectionFailure("remote_link_unreachable") }
      throw RemoteConnection.mappedTlsFailure(failure)
    }
    connection.stateUpdateHandler = nil
  }

  // MARK: - ByteInput / ByteOutput (블로킹)

  public func read(_ maxLength: Int) throws -> [UInt8] {
    if maxLength <= 0 { return [] }
    let done = DispatchSemaphore(value: 0)
    var result: Result<[UInt8], Error> = .success([])
    connection.receive(minimumIncompleteLength: 1, maximumLength: maxLength) { data, _, isComplete, error in
      if let error {
        result = .failure(ConnectionFailure("remote_link_unreachable"))
        _ = error
      } else if let data, !data.isEmpty {
        result = .success([UInt8](data))
      } else if isComplete {
        result = .success([]) // EOF
      } else {
        result = .success([])
      }
      done.signal()
    }
    let timeout: DispatchTime = readTimeoutMs.map { .now() + .milliseconds($0) } ?? .distantFuture
    if done.wait(timeout: timeout) == .timedOut {
      connection.cancel()
      throw ConnectionFailure("remote_link_unreachable")
    }
    return try result.get()
  }

  public func setReadTimeout(milliseconds: Int?) {
    readTimeoutMs = milliseconds
  }

  public func write(_ bytes: [UInt8]) throws {
    let done = DispatchSemaphore(value: 0)
    var failure: Error?
    connection.send(content: Data(bytes), completion: .contentProcessed { error in
      failure = error
      done.signal()
    })
    if done.wait(timeout: .now() + .milliseconds(RemoteConnection.timeoutMs)) == .timedOut {
      connection.cancel()
      throw ConnectionFailure("remote_link_unreachable")
    }
    if failure != nil { throw ConnectionFailure("remote_link_unreachable") }
  }

  public func flush() throws {
    // NWConnection.send는 버퍼링 없이 전달을 큐잉한다 — 별도 flush 불필요.
  }

  public func close() {
    connection.cancel()
  }
}
